src/uct/ib/rc/base/rc_ep.h

```diff
@@ -151,4 +203,85 @@ uct_rc_ep_tx_posted(uct_rc_ep_t *ep, int signaled)
     }
 }
 
+static UCS_F_ALWAYS_INLINE uint8_t
+uct_rc_ep_get_fc_hdr(uint8_t id)
+{
+    return (id & UCT_RC_EP_FC_MASK);
+}
+
+static UCS_F_ALWAYS_INLINE uint8_t
+uct_rc_ep_fc_is_real_am(uint8_t id)
+{
+    return ((id & UCT_RC_EP_FC_MASK) ^ UCT_RC_EP_FC_PURE_GRANT);
+}
+
+static UCS_F_ALWAYS_INLINE void
+uct_rc_ep_update_fc(uct_rc_iface_t *iface, uct_rc_ep_t *ep)
+{
+    uint8_t flags = (ep->fc_wnd == iface->config.fc_hreq_thld) ?
+                     UCT_RC_EP_FC_FLAG_HREQ :
+                    (ep->fc_wnd == iface->config.fc_sreq_thld) ?
+                     UCT_RC_EP_FC_FLAG_SREQ : 0;
+
+    /* Do not update wnd if we have just sent special FC AM with credits grant. 
+     * It can't be bundled with any of FC requests, so we may miss some of the 
+     * thresholds in his case. */
+    if (ucs_likely(uct_rc_ep_fc_is_real_am(ep->flags))) {
+        ep->fc_wnd--;
+    }
+
+    /* Update ep FC bits */
+    ep->flags = flags;
+}
+
+static UCS_F_ALWAYS_INLINE void
+uct_rc_ep_handle_fc(uct_rc_iface_t *iface, uint32_t qp_num, uint8_t fc_hdr)
+{
+    uct_rc_ep_t  *ep = uct_rc_iface_lookup_ep(iface, qp_num);
+    ucs_status_t status;
+    int16_t      batch;
+
+    if (fc_hdr & UCT_RC_EP_FC_FLAG_GRANT) {
+        /* Got either grant flag or special FC grant message. To preserve
+         * ordering we have to dispatch all pending operations if current
+         * wnd is <= 0 (otherwise it will be dispatched by tx progress). Limit
+         * the number of elements to be dispatched by the lowest available ep
+         * resource.
+         * TODO: Make it more fair (introduce reasonable limit for batch).
+         *       Would be good to distinguish AM somehow. */
+        batch = (ep->fc_wnd <= 0) ?
+                ucs_min(iface->config.fc_wnd_size, ep->available) : 0;
+
+        /* Peer granted resources, so update wnd */
+        ep->fc_wnd = iface->config.fc_wnd_size;
+
+        if (batch) {
+            ucs_arbiter_group_schedule(&iface->tx.arbiter, &ep->arb_group);
```

Comment:
why do we need to limit the arbiter to 'batch', it would sends as much as possible anyway